### PR TITLE
Add Lego Rock Raiders

### PR DESCRIPTION
**Pull request by Zemogiter**
This game requires a d3drm.dll file inside it's directory or else it will crash on loading bar before main menu. However I get an error when the script attempts to download said file:

```
13:09:08.506 [ERROR] org.phoenicis.javafx.views.common.ErrorMessage (l.75) - org.phoenicis.tools.http.DownloadException: Download of http://s2.pliki.info/5709/d3drm.dll has failed
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:74)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:58)
	at jdk.nashorn.internal.scripts.Script$Recompilation$369$1124$\^eval\_.Downloader#get([Utils, Functions, Net, Download]:57)
	at jdk.nashorn.internal.scripts.Script$Recompilation$361$508AA$\^eval\_.L:17(<eval>:19)
	at jdk.nashorn.internal.scripts.Script$Recompilation$310$560$\^eval\_.go([Engines, Wine, QuickScript, InstallerScript]:94)
	at jdk.nashorn.internal.scripts.Script$289$\^eval\_.:program(<eval>:8)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.FileNotFoundException: /home/jonasz/.Phoenicis/containers/wineprefix/Lego Rock Raiders/drive_c/Program Files/LEGO Media/Games/Rock Raiders (Is a directory)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:72)
	... 19 more
```

### CODE REVIEW

**Comment by plata** [comment]
I think the `Downloader` expects a file name and not a directory name.

**Comment by plata** [comment]
By the way: we have updated our naming conventions for pull requests. Check the README.

**Comment by Zemogiter** [comment]
The game now loads but cutscenes will not load unless some really old codecs called Indeo are installed. I'm currently in process of porting the bash script from POL4 to POL5's javascript format.

**Comment by plata** [comment]
Ok. Let me know when this is ready for review.

**Comment by Zemogiter** [comment]
I've tried to add 
```
        var SetupFile = new Downloader()
            .wizard(wizard)
            .url("http://ftp.us.dell.com/audio/IV5PLAY.EXE")
            .checksum("809f55efb04b029b71992f17dc160b361298bfdd")
            .to(wine.prefixDirectory + "drive_c/IV5PLAY.EXE")
            .get();
        this.run(SetupFile, [SetupFile, "/q", "/c:\"IV5PLPAY.EXE /q\""])
        .wait(tr("Please wait while {0} is installed ...", "Indeo Codecs 5"));
        var registrySettings = new AppResource().application([TYPE_ID, CATEGORY_ID, APPLICATION_ID]).get("fix.reg");
        wine.regedit().patch(registrySettings);
```
in .preInstall(wine,wizard) instruction to skip having to make an entire verb. Sadly I've ended up with this error:
```
org.phoenicis.javafx.views.common.ErrorMessage (l.75) - javax.script.ScriptException: TypeError: this.run is not a function in <eval> at line number 24
	at jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(NashornScriptEngine.java:470)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:454)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: <eval>:24 TypeError: this.run is not a function
	at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:213)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:185)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:172)
	at jdk.nashorn.internal.runtime.Undefined.lookup(Undefined.java:102)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:106)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:98)
	at jdk.internal.dynalink.support.CompositeTypeBasedGuardingDynamicLinker.getGuardedInvocation(CompositeTypeBasedGuardingDynamicLinker.java:176)
	at jdk.internal.dynalink.support.CompositeGuardingDynamicLinker.getGuardedInvocation(CompositeGuardingDynamicLinker.java:124)
	at jdk.internal.dynalink.support.LinkerServicesImpl.getGuardedInvocation(LinkerServicesImpl.java:154)
	at jdk.internal.dynalink.DynamicLinker.relink(DynamicLinker.java:253)
	at jdk.nashorn.internal.scripts.Script$Recompilation$294$473AA$\^eval\_.L:15(<eval>:24)
	at jdk.nashorn.internal.scripts.Script$Recompilation$259$560$\^eval\_.go([Engines, Wine, QuickScript, InstallerScript]:79)
	at jdk.nashorn.internal.scripts.Script$238$\^eval\_.:program(<eval>:9)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	... 10 more
```
Which indicates the this.run line is probably only for verbs. I am not sure if I can make that verb after all.

**Comment by madoar** [comment]
@Zemogiter `run` is a method of `wine`, see [here](https://github.com/PhoenicisOrg/Scripts/blob/master/Engines/Wine/Engine/Object/script.js#L197) for the definition.
So it should work if you replace `this.run` with `wine.run`, because `wine` is passed to the `preInstall` hook.

**Comment by Zemogiter** [comment]
@madoar tried that and now I get this error instead:
```
javax.script.ScriptException: TypeError: Cannot read property "split" from undefined in [Engines, Wine, Engine, Object] at line number 203
	at jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(NashornScriptEngine.java:470)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:454)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: [Engines, Wine, Engine, Object]:203 TypeError: Cannot read property "split" from undefined
	at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:213)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:185)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:172)
	at jdk.nashorn.internal.runtime.Undefined.get(Undefined.java:157)
	at jdk.nashorn.internal.scripts.Script$Recompilation$75$4652AAA$\^eval\_.run([Engines, Wine, Engine, Object]:203)
	at jdk.nashorn.internal.scripts.Script$Recompilation$83$490AA$\^eval\_.L:15(<eval>:24)
	at jdk.nashorn.internal.scripts.Script$Recompilation$48$560$\^eval\_.go([Engines, Wine, QuickScript, InstallerScript]:79)
	at jdk.nashorn.internal.scripts.Script$27$\^eval\_.:program(<eval>:9)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	... 10 more
```

**Comment by Zemogiter** [comment]
I've replaced `wine.run(SetupFile, [SetupFile, "/q", "/c:\"IV5PLPAY.EXE /q\""])` with `wine.run("IV5PLPAY.EXE /q")` and now the script works. But it did not help with the cutscenes. They still do not play. I'm aware of two Indeo Codecs 5 .exe files - the one used in POL4 script which is also the one I was using. The other one is http://s3.amazonaws.com/moviecodec/files/iv5setup.exe which is listed in the game's page on WineHQ.

**Comment by Zemogiter** [comment]
Seems like codacy still complaind about `new Download()` function.

**Comment by plata** [comment]
Try to merge current master to fix Codacy.

**Comment by Zemogiter** [comment]
Script is ready and waiting for review.

File: Applications/Games/Lego Rock Raiders/Local/script.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+include(["Engines", "Wine", "QuickScript", "LocalInstallerScript"]);
+include(["Utils", "Functions", "Filesystem", "Files"]);
+//include(["Utils", "Functions", "Apps", "Resources"]);
```
Comment (plata) [inline Applications/Games/Lego Rock Raiders/Local/script.js]: Please remove unused code.

File: Applications/Games/Lego Rock Raiders/application.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+  	"name":	"Lego Rock Raiders",
+  	"description": "A Lego Sci-Fi Real time strategy game. The object is to complete missions by mining a certain quota of energy crystal, while defending your mining base or even struggling to keep a breathable atmosphere in the mine. "
```
Comment (plata) [inline Applications/Games/Lego Rock Raiders/application.json]: Remove space at end of description.

**Comment by plata** [comment]
@madoar @qparis I'm not sure about the file which is downloaded. Do we have to consider this as `requiresPatch`. Could it be illegal in some countries? If so, can we even merge this?

**Comment by madoar** [comment]
I don't think that this can be considered as `requiresPatch`.
As far as I understand `requiresPatch`, this optional means that the installation files need to be modified afterwards, to work correctly, which is not the case as far as I understand.
I'm not sure if the installation and download `d3drm` is illegal in some countries, I guess it depends on who the owner of the file is and under which license it is available.

**Comment by plata** [comment]
From the wiki:

> Application cannot work without patching the original protection. It may be illegal to install it depending on your country.

To me the downloaded file looks a bit like this might be the case here.

**Comment by Zemogiter** [comment]
Ready for review.

**Comment by plata** [comment]
I'm still unsure about the patch. I would like to hear @qparis opinion about it.

**Comment by Zemogiter** [comment]
Upon downloading the .dll file I get this error message:
```
org.phoenicis.tools.http.DownloadException: Download of http://s2.pliki.info/5709/d3drm.dll has failed
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:74)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:58)
	at jdk.nashorn.internal.scripts.Script$Recompilation$212$2364$\^eval\_.get([utils, functions, net, download]:113)
	at jdk.nashorn.internal.scripts.Script$Recompilation$205$424AA$\^eval\_.L:14(<eval>:16)
	at jdk.nashorn.internal.scripts.Script$Recompilation$160$565$\^eval\_.go([engines, wine, quick_script, installer_script]:89)
	at jdk.nashorn.internal.scripts.Script$126$\^eval\_.:program(<eval>:8)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.FileNotFoundException: function () {
    return this._implementation.getContainerDirectory(this._implementation.getWorkingContainer());
}drive_c/Program Files/LEGO Media/Games/Rock Raiders/d3drm.dll (Nie ma takiego pliku ani katalogu)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at org.phoenicis.tools.http.Downloader.get(Downloader.java:72)
	... 19 more
```

**Comment by plata** [comment]
Look at the output:
```
Caused by: java.io.FileNotFoundException: function () {
    return this._implementation.getContainerDirectory(this._implementation.getWorkingContainer());
}drive_c/Program Files/LEGO Media/Games/Rock Raiders/d3drm.dll
```
This is caused by:
```Javascript
var GameDir = wine.prefixDirectory + "drive_c/" + wine.programFiles() + "/LEGO Media/Games/Rock Raiders/d3drm.dll";
```
You're missing `()` behind `wine.prefixDirectory` so it copied the function there instead of executing it.

**Comment by Zemogiter** [comment]
Its weird because this part of the script worked the last time it was tested (April 2018). Anyway uploading the fixed version.

**Comment by Zemogiter** [comment]
After the fix the error message changed. This time it does not refer to anything inside the script but rather something called eval
The error happens right after the installation of the game and before the .dll file is downloaded.
```
javax.script.ScriptException: ReferenceError: "Installer" is not defined in <eval> at line number 25
	at jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(NashornScriptEngine.java:470)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:454)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: <eval>:25 ReferenceError: "Installer" is not defined
	at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at jdk.nashorn.internal.runtime.ECMAErrors.referenceError(ECMAErrors.java:319)
	at jdk.nashorn.internal.runtime.ECMAErrors.referenceError(ECMAErrors.java:291)
	at jdk.nashorn.internal.objects.Global.__noSuchProperty__(Global.java:1441)
	at jdk.nashorn.internal.scripts.Script$865$\^eval\_.:program(<eval>:25)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	... 10 more
```

**Comment by plata** [comment]
Yeah, that's because we've changed the scripts a bit such that the interface to the software becomes clearer. Check some other script on master to figure out how it should look like.

**Comment by Zemogiter** [comment]
The current state of this script is as follows:

1. The script installs the game from local source.
2. Then it downloads the d3drm.dll file inside the game directory.
3. Now it applies the fix for cutscenes. And here is where I have problems. It downloads the zip file with an .exe file for codec and an .dll file to make cutscenes work. Extracting works but I don't know how to run that exe file after it's extracted. Next is how to register the .dll file? It's needed because just runing the iv50 codecs installation isn't enough. I know how to do it in cmd (regsvr32 ir50_32.dll) but not in javascript.

**Comment by plata** [comment]
To install the exe, use `wine.run`. There's a plugin for [regsvr32](https://github.com/PhoenicisOrg/scripts/blob/69c90d5efd842385a0d29f6b451298087872bc2c/Engines/Wine/Plugins/regsvr32/script.js).

**Comment by Zemogiter** [comment]
After the changes I get this error:
```
java.lang.IllegalMonitorStateException
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at jdk.nashorn.internal.scripts.Script$Recompilation$8349$903AA$\^eval\_.run#L:24(<eval>:46)
	at jdk.nashorn.internal.scripts.Script$Recompilation$8290$565$\^eval\_.go([engines, wine, quick_script, installer_script]:90)
	at jdk.nashorn.internal.scripts.Script$Recompilation$8279$557$\^eval\_.run(<eval>:15)
	at org.phoenicis.scripts.Installer$$NashornJavaAdapter.run(Unknown Source)
	at jdk.nashorn.internal.scripts.Script$8251$\^eval\_.:program(<eval>:61)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
Strangely, this error didn't stop the iv50 codec installer to show up.

**Comment by plata** [comment]
I think you have a problem with `wine.wait`. This should be fixed as soon as my SteamScript fix is merged.

**Comment by Zemogiter** [comment]
Strangely enough, once I remove the wait() instruction I get an error regarding the regsvr32:
```
javax.script.ScriptException: TypeError: this.regsvr32 is not a function in <eval> at line number 52
	at jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(NashornScriptEngine.java:470)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:454)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: <eval>:52 TypeError: this.regsvr32 is not a function
	at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:213)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:185)
	at jdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:172)
	at jdk.nashorn.internal.runtime.Undefined.lookup(Undefined.java:102)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:106)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:98)
	at jdk.internal.dynalink.support.CompositeTypeBasedGuardingDynamicLinker.getGuardedInvocation(CompositeTypeBasedGuardingDynamicLinker.java:176)
	at jdk.internal.dynalink.support.CompositeGuardingDynamicLinker.getGuardedInvocation(CompositeGuardingDynamicLinker.java:124)
	at jdk.internal.dynalink.support.LinkerServicesImpl.getGuardedInvocation(LinkerServicesImpl.java:154)
	at jdk.internal.dynalink.DynamicLinker.relink(DynamicLinker.java:253)
	at jdk.nashorn.internal.scripts.Script$Recompilation$700$903AA$\^eval\_.run#L:24(<eval>:52)
	at jdk.nashorn.internal.scripts.Script$Recompilation$642$565$\^eval\_.go([engines, wine, quick_script, installer_script]:89)
	at jdk.nashorn.internal.scripts.Script$Recompilation$631$557$\^eval\_.run(<eval>:15)
	at org.phoenicis.scripts.Installer$$NashornJavaAdapter.run(Unknown Source)
	at jdk.nashorn.internal.scripts.Script$585$\^eval\_.:program(<eval>:61)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	... 10 more
```

**Comment by plata** [comment]
Same issue like before: `wine.regsvr32()`

**Comment by Zemogiter** [comment]
My local repo is up to date and I still get the `java.lang.IllegalMonitorStateException` error from before when using the `.wait()` after `wine.run`

**Comment by plata** [comment]
Try to split
```Javascript
wine.run(wine.prefixDirectory() + "/drive_c/RockRaidersCodec/iv5setup.exe").wait();
```
to
```Javascript
wine.run(wine.prefixDirectory() + "/drive_c/RockRaidersCodec/iv5setup.exe");
wine.wait();
```

**Comment by Zemogiter** [comment]
Ready for review.

File: Applications/Games/Lego Rock Raiders/Local/script.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            .name("Lego Rock Raiders")
+            .editor("LEGO Media")
+            .author("Zemogiter")
+            .category("Games")
+            .executable("LegoRR.exe")
+            .preInstall(function(wine) {
+                wine.quartz();
+            })
+            .postInstall(function(wine,wizard) {
+                var GameDir = wine.prefixDirectory() + "drive_c/" + wine.programFiles() + "/LEGO Media/Games/Rock Raiders/d3drm.dll";
```
Comment (plata) [inline Applications/Games/Lego Rock Raiders/Local/script.js]: variable names should be camel case: `gameDir`

Apart from this, the variable does not contain the game dir but the dll path, so please use a matching name.

File: Applications/Games/Lego Rock Raiders/Local/script.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                new Downloader()
+                .wizard(wizard)
+                .url("http://s2.pliki.info/5709/d3drm.dll")
+                .checksum("dde9e3b8c264957ae0a017d371293123")
+                .algorithm("MD5")
+                .to(GameDir)
+                .get();
+                var DiskCRoot = wine.prefixDirectory() + "drive_c/RockRaidersCodec_490085.zip";
```
Comment (plata) [inline Applications/Games/Lego Rock Raiders/Local/script.js]: see above

File: Applications/Games/Lego Rock Raiders/Local/script.js (hunk 3 above)
Comment (Zemogiter) [reply Applications/Games/Lego Rock Raiders/Local/script.js]: But the script works no matter if variable names start with capital letter or not.

**Comment by plata** [comment]
The names are still not good (e.g. you have a dll file and call it `dllDir`).

File: Applications/Games/Lego Rock Raiders/application.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+    "name"        : "Lego Rock Raiders",
+    "id"          : "lego_rock_raiders",
+    "description" : "A Lego Sci-Fi Real time strategy game. The object is to complete missions by mining a certain quota of energy crystal, while defending your mining base or even struggling to keep a breathable atmosphere in the mine."
```
Comment (plata) [inline Applications/Games/Lego Rock Raiders/application.json]: Shouldn't "Real" be lower case? "object" or "objective"?

File: Applications/Games/Lego Rock Raiders/Local/script.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+    "scriptName"                 : "Local",
+    "id"                         : "lego_rock_raiders",
+    "compatibleOperatingSystems" : [
+        "MACOSX",
+        "LINUX"
+    ],
+    "testingOperatingSystems" : [
+        "MACOSX",
```
Comment (plata) [inline Applications/Games/Lego Rock Raiders/Local/script.json]: Is it still correct that this is testing?

Comment (Zemogiter) [reply Applications/Games/Lego Rock Raiders/Local/script.json]: Yes.

**Comment by Zemogiter** [comment]
There is a long-runing issue when installing Indeo codecs. At 33% installation (Indeo Video 5 Progresive Download DirectShow filter) it displays the following error:
```
the file C:\windows\system32\IvfSrc.ax failed to self-register. This could be caused by corrupt registry or corrupt or impropetly installed DirectX Media runtime libs
```
Seems lie this error is wine-related and looking for solution I found this:
https://bugs.winehq.org/show_bug.cgi?id=11419
the bug was abandoned :(

**Comment by qparis** [comment]
Sorry for the delay ; what is the status if this MR? Do you need some review from me?

**Comment by Zemogiter** [comment]
The only issue with this game is lack of cutscenes. I've tried to follow the guidelines at AppDB and install Indeo 5 codecs but even when I unchecked the two options from the codec's installer that were causing the setup to fail (Indeo Video 5 Progresive Download DirectShow filter and Indeo 5 Netscape Browser Extension) the installation of those codecs did nothing. I think due to this came being DirectX 6 makes it impossible to get it to run fully on modern machines. Tried many ways to "modernize" the code such as Cafeteria mod but it either didn't help anything or screwed my display's resolution. @plata wanted your opinion on the patch. Lego Rock Raiders uses an obsolete method of copy protection called SafeDisc. This .dll patch is needed for both Wine and Windows since Microsoft removed SafeDisc support somewhere between XP and win7 due to security concerns.
However a new report was posted on AppDB with wine version 3.0.3 being used. I will try to integrate the guidelines from that report into the script and see if this works.

**Comment by qparis** [comment]
Ok so if we follow POLv4 policy, we should not provide Safedisc / securerom or any copy protection workaround related patch.

Instead, we clearly explain this dilemma to the end user, why is the patch required and that using them is not legal on many circonstances. 

Then, it’s up to them to take the responsibility on finding it and applying it, denpending on the législation of where they come from

**Comment by Zemogiter** [comment]
@qparis we already kinda said that to the end user in https://phoenicisorg.github.io/scripts/Develop/script-json/

**Comment by qparis** [comment]
Yep that just for the legal aspect. The difference is that we (POL) should not take the responsibility of distributing (or providing links to) these patches

**Comment by Zemogiter** [comment]
Wow I just did it! Cutscenes are working! So basically I've followed this:
https://appdb.winehq.org/objectManager.php?sClass=version&iId=11356
up to the point where it said ro rename/remove cutscene files. Here I just ran the Indeo codecs setup file and unchecked those two options.

**Comment by plata** [comment]
So blocked until the verbs are implemented?

**Comment by Zemogiter** [comment]
Yes and when making amstream verb earlier today I've discovered our regsvr32 plugin only do the 32-bit variant and if run on a 64bit wineprefix, amstream will require regsvr64 (look at https://github.com/Winetricks/winetricks/blob/master/src/winetricks )

**Comment by plata** [comment]
I was only saying that so that we do not forget about the verbs which are required and the necessary changes for `regsvr32/64`. If you can fix it directly, that's of course possible as well.

**Comment by Zemogiter** [comment]
The script is now fully working and can be merged.

**Comment by plata** [comment]
As the patch is removed, I think we're good.

**Pull request merged**